File: 326/old/cpp/etude1.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <deque>
#include <list>
using namespace std;

class Coin{
    int value;
    public:
        Coin(int coinValue){
            value = coinValue;
        }
        int returnValue(){
            return value;
        }
};

queue<int> swapQueue;

void Swap(vector<Coin> c){
    Coin firstCoin = c[swapQueue.front()];
    Coin secondCoin = c[swapQueue.front() + 1];
    Coin handOff = c[swapQueue.front() + 1];
    secondCoin = firstCoin;
    firstCoin = handOff;
    c[swapQueue.front()] = firstCoin;
    c[swapQueue.front() + 1] = secondCoin;
    swapQueue.pop();
}

void Sweep(vector<Coin> c){
    for(int i = 0; i < c.size(); i++){
        if(!(c[i].returnValue() == c[i + 1].returnValue())){
            swapQueue.push(i);
            i++;
        }
    }
}

bool checkHeadsSequence(vector<Coin> c){
    cout << "checkHeadsSequence...";
    int breakPointHeads = 0;
    bool checkHeads = false;
    vector<int> headsSequence;
    for(int u = 0; u < c.size(); u++){
        headsSequence.push_back(c[u].returnValue());
        u++;
    }
    for(int y = 0; y < headsSequence.size(); y++){
        if(headsSequence[0] == headsSequence[y]){
            checkHeads = true;
        } else{
            breakPointHeads = y;
            return false;
        }
    }
    return checkHeads;
}

bool checkTailsSequence(vector<Coin> c){
    cout << "checkTailsSequence...";
    int breakPointTails = 0;
    bool checkTails = false;
    vector<int> tailsSequence;
    for(int u = 0; u < c.size(); u++){
        tailsSequence.push_back(c[u].returnValue());
        u++;
    }
    for(int y = 0; y < tailsSequence.size(); y++){
        if(tailsSequence[0] == tailsSequence[y]){
            checkTails = true;
        } else{
            breakPointTails = y;
            return false;
        }
    }
    return checkTails;
}

vector<Coin> Evaluate(vector<Coin> c){
    Sweep(c);
    bool isAlternating = false;
    while(!isAlternating){
        Swap(c);
        if(swapQueue.size() == 0){
            bool tailsCheck = checkTailsSequence(c);
            bool headsCheck = checkHeadsSequence(c);
            if(tailsCheck && headsCheck){
                isAlternating = true;
            } else{
                Sweep(c);
            }
        }
    }
    return c;
}

int main(){
    int count = 0;
    vector<int> totalMoveCountVector;
    vector<int> indexesVector;
    vector<Coin> coinVector;
    cout << "1 for input\n2 to print 2 < n < 20\n";
    int choice;
    cin >> choice;
    switch(choice){
        case 1:
            count = 0;
            coinVector.empty();
            cout << "Enter heads\n";
            int headsCount;
            cin >> headsCount;
            int tailsCount = headsCount;
            for(int k = 0; k < headsCount; k++){
                Coin c(1);
                coinVector.push_back(c);
            }
            for(int p = 0; p < tailsCount; p++){
                Coin c(0);
                coinVector.push_back(c);
            }
            Evaluate(coinVector);
            break;
    }
    return EXIT_SUCCESS;
}

File: 326/old/etude1/etude1.java
import java.util.*;
public class etude1{
    private static Queue<Integer> Queue = new LinkedList<Integer>();
    private static Scanner sc = new Scanner(System.in);
    static Integer Count = 0;
    static Integer BreakPointHeads = 0;
    static Integer BreakPointTails = 0;
    enum Classification{
        H,
        T
    }
    public static void main(String[] args){
        ArrayList<Coin> CoinsList = new ArrayList<Coin>();
        ArrayList<Integer> TotalMoveCountsList = new ArrayList<Integer>();
        ArrayList<Integer> IndexesList = new ArrayList<Integer>();
        System.out.println("1 to enter input");
        System.out.println("2 to print 2 < n < 20");
        Integer Choice = sc.nextInt();
        switch(Choice){
            case 1:
                Count = 0;
                CoinsList.clear();
                System.out.println("Enter heads");
                Integer HeadsCount = sc.nextInt();
                Integer TailsCount = HeadsCount;
                for(int k = 0; k < HeadsCount; k++){
                    CoinsList.add(new Coin(1));
                }
                for(int p = 0 ; p < TailsCount; p++){
                    CoinsList.add(new Coin(0));
                }
                Evaluate(CoinsList);
                System.out.println("\nMove Count: " + Count);
                break;
            case 2:
                for(int i = 2; i < 20; i++){
                    CoinsList.clear();
                    Count = 0;
                    for(int u = 0; u < i; u++){
                        CoinsList.add(new Coin(0));
                    }
                    for(int j = 0; j < i; j++){
                        CoinsList.add(new Coin(1));
                    }
                    Evaluate(CoinsList);
                    TotalMoveCountsList.add(Count);
                    IndexesList.add(i);
                }
                System.out.println("\nTotal move counts: ");
                for(int i = 0; i < TotalMoveCountsList.size(); i++){
                    System.out.println("n: " + IndexesList.get(i) + "\tcount: " + TotalMoveCountsList.get(i));
                }
                break;
        }
    }

    public static void Sweep(ArrayList<Coin> Coins){
        for(int i = 0; i < Coins.size() - 1; i++){   
            if(!(Coins.get(i).value.equals(Coins.get(i + 1).value))){      
                    Queue.add(i);
                    i++;                                                                                             
            } 
        } 
    }
    private static void Swap(ArrayList<Coin> Coins){
        Coin First = Coins.get(Queue.peek());
        Coin Second = Coins.get(Queue.peek() + 1);
        Coin Handoff = Coins.get(Queue.peek() + 1);
        Second = First;
        First = Handoff;
        Coins.set(Queue.peek(), First);
        Coins.set(Queue.peek() + 1, Second);
        Queue.remove(); 
        Count++; 
        for(Coin c : Coins){
            if(c.value.equals(0)) System.out.print(Classification.T + " ");
            if(c.value.equals(1)) System.out.print(Classification.H + " ");
        } 
        System.out.println();
    }

    public static ArrayList<Coin> Evaluate(ArrayList<Coin> Coins){
        Sweep(Coins);
        System.out.println();
        Boolean IsAlternating = false;
        while(!IsAlternating){    
            Swap(Coins);
            if(Queue.isEmpty()){
                Boolean TailsCheck = (CheckTailsSequence(Coins) ? true : false);
                Boolean HeadsCheck = (CheckHeadsSequence(Coins) ? true : false);   
                if(TailsCheck && HeadsCheck){
                    IsAlternating = true;
                    
                } else{               
                    Sweep(Coins);
                }
            }  
        }
        return Coins;
    }
    
    private static Boolean CheckHeadsSequence(ArrayList<Coin> Coins){
        BreakPointHeads = 0;
        Boolean CheckHeads = false;
        ArrayList<Integer> HeadsSequence = new ArrayList<Integer>();               
        for(int u = 0; u < Coins.size(); u++){
            HeadsSequence.add(Coins.get(u).value);
            u++;
        }
        for(int k = 0; k < HeadsSequence.size(); k++){
            if(HeadsSequence.get(0).equals(HeadsSequence.get(k))){
                CheckHeads = true;
            } else{
                BreakPointHeads = k;
                return false;
            }
        }
        return CheckHeads;
    }

    private static Boolean CheckTailsSequence(ArrayList<Coin> Coins){
        BreakPointTails = 0;
        ArrayList<Integer> TailsSequence = new ArrayList<Integer>();
        Boolean CheckTails = false;           
        for(int u = 1; u < Coins.size(); u++){
            TailsSequence.add(Coins.get(u).value);
            u++;
        }
        for(int i = 0; i < TailsSequence.size(); i++){
            if(TailsSequence.get(0).equals(TailsSequence.get(i))){
                CheckTails = true;
            } else{
                BreakPointTails = i;
                return false;
            }
        }
        return CheckTails;
    }  
    private static class Coin{
        public Integer value;
        public Coin(Integer coinValue){
            this.value = coinValue;
        }
    }
}

